File: code/word-count.cpp
// Read from stdin and produce a sorted word frequency table

// Compile:
// g++ -o word_count -std=c++11 word_count.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <map>
#include <cstdio>

using namespace std;


// Count unique words from input text stream and print word frequency table
void word_count(istream& stream) {
	string line;
	map<string, size_t> word_counts;

	// process text from input stream
	while (getline(stream, line)) {

		// convert to lowercase, extract words, and count unique words
		size_t start = 0;
		size_t line_length = line.length();
		for (size_t i = 0; i <= line_length; ++i) {
			// look for word boundary as marked by non-alphabet characters or EOL
			if (i == line_length || !isalpha(line[i])) {
				// met word boundary: extract word
				// skip empty words
				if (i > start) {
					string word = line.substr(start, i - start);
					if (word_counts.find(word) == word_counts.end()) {
						// word does not exist: add it
						word_counts[word] = 1;
					} else {
						// word exists: increment it
						++word_counts[word];
					}
				}
				// shift to next word
				start = i+1;
			} else if (i < line_length) {
				// convert all alphabet characters to lowercase
				line[i] = ::tolower(line[i]);
			}
		}
	}

	// sort words by frequency
	vector<string> words;
	words.reserve(word_counts.size());
	for (auto it = word_counts.begin(); it != word_counts.end(); ++it) {
		words.push_back(it->first);
	}
	auto compare_words = [&word_counts](const string& a, const string& b) {
		return word_counts[b] < word_counts[a]
			|| (word_counts[b] == word_counts[a] && a < b);
	};
	sort(words.begin(), words.end(), compare_words);

	// print word frequency table
	for (auto it = words.begin(); it != words.end(); ++it) {
		printf("%4lu %s\n", word_counts[*it], it->c_str());
	}
}

int main(int argc, char* argv[]) {

	// process input stream or given text file
	if (argc == 1) {
		word_count(cin);
	} else {
		ifstream in(argv[1]);
		word_count(in);
	}

	return 0;
}

File: code/word-count.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func isAlpha(c byte) bool {
	return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z')
}

func toLower(c byte) byte {
	if 'A' <= c && c <= 'Z' {
		return c + 'a' - 'A'
	} else {
		return c
	}
}

// StringCompare is the type of a "Less" function that defines the ordering
type StringCompare func(x, y string) bool

func (cmp StringCompare) Sort(strings []string) {
	sorter := &stringSorter{
		strings: strings,
		compare: cmp,
	}
	sort.Sort(sorter)
}

type stringSorter struct {
	strings []string
	compare StringCompare
}

func (s *stringSorter) Len() int {
	return len(s.strings)
}

func (s *stringSorter) Swap(i, j int) {
	s.strings[i], s.strings[j] = s.strings[j], s.strings[i]
}

func (s *stringSorter) Less(i, j int) bool {
	return s.compare(s.strings[i], s.strings[j])
}

func main() {
	var words []string

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		token := scanner.Bytes()
		start := 0
		n := len(token)
		for i := 0; i <= n; i++ {
			// non-alphabeta character and end of line demarcate word boundary
			if i == n || !isAlpha(token[i]) {
				if start < i {
					// extract word
					words = append(words, string(token[start:i]))
				}
				// shift to next word
				start = i + 1
			} else if i < n {
				// do simultaneous lowercase conversion during loop
				token[i] = toLower(token[i])
			}
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "Error while reading stdin: ", err)
		os.Exit(1)
	}

	// count unique words
	word_counts := make(map[string]int)
	for _, w := range words {
		word_counts[w] += 1
	}

	var words_unique []string
	for k := range word_counts {
		words_unique = append(words_unique, k)
	}

	// closure that orders the word
	word_order := func(x, y string) (less bool) {
		if word_counts[x] == word_counts[y] {
			// lexicographical order
			less = x < y
		} else {
			// descending word frequency
			less = word_counts[x] > word_counts[y]
		}
		return
	}

	// sort the unique words
	StringCompare(word_order).Sort(words_unique)

	// print word frequency table
	for _, w := range words_unique {
		fmt.Printf("%4d %s\n", word_counts[w], w)
	}

}
